oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/inference/InferenceService.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.index.elastic.query.inference;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * EXPERIMENTAL: A service that sends text to an inference service and receives embeddings in return.
+ * The embeddings are cached to avoid repeated calls to the inference service.
+ */
+public class InferenceService {
+
+    private final URL url;
+    private final Cache<String, List<Float>> cache;
+
+    public InferenceService(String url, int cacheSize) {
+        try {
+            this.url = new URL(url);
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException("Invalid URL: " + url, e);
+        }
+        this.cache = new Cache<>(cacheSize);
+    }
+
+    public List<Float> embeddings(String text, int timeoutMillis) {
+        if (cache.containsKey(text)) {
+            return cache.get(text);
+        }
+
+        HttpURLConnection connection;
```

Comment:
Not thread safe. Should it be?
And if it should be thread safe, this does not protect against 2 or more threads calling this method for the same text at the same time, it will make the HTTP call several times for the same value.

Reply:
This part requires more work. We can improve caching when the inference support will be not experimental anymore. For now, I will leave it like this. In the worst scenario, we will call the external service twice (or more) with the same input but the functionality is not affected.